sklearn/linear_model/ridge.py

```diff
@@ -995,7 +1015,131 @@ def _diag_dot(self, D, B):
+        X_mX_m = np.empty((n_samples, n_samples), dtype=X.dtype)
+        X_mX_m[:, :] = np.dot(X_mean, X_mean)
+        X_mX_m *= self._sqrt_sw
+        X_mX_m *= self._sqrt_sw[:, None]
+        return (safe_sparse_dot(X, X.T, dense_output=True) + X_mX_m
+                - X_mX - X_mX.T, X_mean)
+
+    def _compute_covariance(self, X, center=True):
+        """Computes centered covariance matrix.
+
+        If ``center`` is ``True``, compute
+        (X - X.mean(axis=0)).T.dot(X - X.mean(axis=0))
+        else
+        X.T.dot(X)
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input uncentered data.
+        center : bool, default is True
+            Whether or not to remove the mean from ``X``.
+
+        Returns
+        -------
+        covariance : ndarray, shape (n_features, n_features)
+            The covariance matrix.
+        X_m : ndarray, shape (n_feature,)
+            The mean of ``X`` for each feature.
+        """
+        if not center:
+            # in this case centering has been done in preprocessing
+            # or we are not fitting an intercept.
+            X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X.T, X, dense_output=True), X_mean
+        # otherwise X is always sparse
+        n_samples = X.shape[0]
+        X_weighted = self._sqrt_sw_matrix.dot(X)
+        X_mean, _ = mean_variance_axis(X_weighted, axis=0)
+        X_mean = X_mean * n_samples / self._weight_sum
+        return (safe_sparse_dot(X.T, X, dense_output=True) -
+                self._weight_sum * np.outer(X_mean, X_mean),
+                X_mean)
+
+    def _sparse_multidot_diag(self, X, A, X_mean):
+        """ compute the diagonal of (X - X_mean).dot(A).dot((X - X_mean).T)
+        without explicitely centering X nor computing X.dot(A)
+        when X is sparse.
+        """
+        intercept_col = self._sqrt_sw
+        scale = self._sqrt_sw
+        batch_size = X.shape[1]
+        diag = np.empty(X.shape[0])
+        for start in range(0, X.shape[0], batch_size):
+            batch = slice(start, min(X.shape[0], start + batch_size), 1)
+            X_batch = np.empty(
+                (X[batch].shape[0], X.shape[1] + self.fit_intercept),
+                dtype=X.dtype
+            )
+            if self.fit_intercept:
+                X_batch[:, :-1] = X[batch].A - X_mean * scale[batch][:, None]
+                X_batch[:, -1] = intercept_col[batch]
+            else:
+                X_batch = X[batch].A
+            diag[batch] = (X_batch.dot(A) * X_batch).sum(axis=1)
+        return diag
+
+    def _set_intercept(self, X_offset, y_offset, X_scale):
+        # add the mean of X which was computed separately if X is sparse
+        if getattr(self, '_X_offset', None) is not None:
+            X_offset += self._X_offset * X_scale
+        super()._set_intercept(X_offset, y_offset, X_scale)
+
+    def _decompose_gram(self, X, y):
```

Comment:
You can put a small docstring (single line) stating for which case we should use it `n_samples` < `n_features`